nrf-hal-common/src/nvmc.rs

```diff
@@ -0,0 +1,101 @@
+//! HAL interface to the Non-Volatile Memory Controller (NVMC) peripheral.
+
+#[cfg(any(feature = "52840"))]
+use crate::pac::NVMC;
+#[cfg(any(feature = "9160"))]
+use crate::pac::NVMC_NS;
+use embedded_storage::nor_flash::{NorFlash, ReadNorFlash};
+
+/// Interface to an NVMC instance.
+pub struct Nvmc<'a, T: Instance, const N: usize> {
+    nvmc: T,
+    storage: &'a mut [u32; N],
+}
+
+impl<'a, T, const N: usize> Nvmc<'a, T, N>
+where
+    T: Instance,
+{
+    /// Takes ownership of the peripheral and storage area.
+    pub fn new(nvmc: T, storage: &'a mut [u32; N]) -> Nvmc<'a, T, N> {
+        Self { nvmc, storage }
+    }
+
+    /// Consumes `self` and returns back the raw peripheral and associated storage.
+    pub fn free(self) -> (T, &'a mut [u32; N]) {
+        (self.nvmc, self.storage)
+    }
+}
+
+impl<'a, T, const N: usize> ReadNorFlash for Nvmc<'a, T, N>
+where
+    T: Instance,
+{
+    type Error = ();
+
+    const READ_SIZE: usize = 4;
+
+    fn try_read(&mut self, _offset: u32, _bytes: &mut [u8]) -> Result<(), Self::Error> {
+        unimplemented!()
+    }
+
+    fn capacity(&self) -> usize {
+        unimplemented!()
+    }
+}
+
+impl<'a, T, const N: usize> NorFlash for Nvmc<'a, T, N>
+where
+    T: Instance,
+{
+    const WRITE_SIZE: usize = 4;
+
+    const ERASE_SIZE: usize = 4 * 1024;
+
+    fn try_erase(&mut self, _from: u32, _to: u32) -> Result<(), Self::Error> {
+        unimplemented!()
+    }
+
+    fn try_write(&mut self, _offset: u32, _bytes: &[u8]) -> Result<(), Self::Error> {
+        unimplemented!()
+    }
+}
+
+pub trait Instance {
+    fn enable_erase();
```

Comment:
I'm not sure if these methods are needed. I'd just do the `#[cfg]`s in the logic above, which imo is much simpler.

Reply:
Commit c71362c35521009a442f727845b6272d8c2a5e63 attends to this.

Reply:
Commit 7958ca2c5be51555b09662955344c06740e33bd9 reverts back to the trait methods because there is no common type representing the NVMC configuration. For the 52840, we have the CONFIG type which is a register of a _CONFIG struct. For the 9160 it is CONFIGNS and _CONFIGNS respectively.

Reply:
Actually, commit e0e44824a62d4bca846db6f2d1b25ff9d98f3764 figures it out. `Deref` FTW.

Reply:
Commit 90b8818617177c76470ed46a4226e4f6461eac7f also adds 52840 considerations.